### PR TITLE
Feature/3.4.2

### PR DESCRIPTION
**Pull request by fabianbl**
Updated to OpenCV 3.4.2 (released on 4.7.2018) and enabled the Aruco module (which got some of the advantages of AprilTags in this PR: https://github.com/opencv/opencv_contrib/pull/1570).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -55,7 +55,7 @@ ExternalProject_Add(opencv3_src
     -DBUILD_opencv_dnn=OFF # Pulls in the system protobuf as a dependency!
     -DBUILD_opencv_dnns_easily_fooled=OFF
     -DBUILD_opencv_cnn_3dobj=OFF
-    -DBUILD_opencv_aruco=OFF
+    -DBUILD_opencv_aruco=ON
```
Comment (eggerk) [inline CMakeLists.txt:58]: will this create a new library? if so, please add it also to the `cs_export` command in #17 after you merge this

File: CMakeLists.txt [hunk 2]
```diff
@@ -82,15 +82,23 @@ ExternalProject_Add(opencv3_src
...
     -DBUILD_opencv_xobjdetect=OFF
-    -DBUILD_opencv_xphoto=OFF .
-  BUILD_COMMAND cd ../${OPENCV_SRC_PATH} && make -j8
-  INSTALL_COMMAND cd ../${OPENCV_SRC_PATH} && make install -j8
+    -DBUILD_opencv_xphoto=OFF ../${OPENCV_SRC_PATH}
+  BUILD_COMMAND cd ../${OPENCV_BUILD_PATH} && make -j8
+  INSTALL_COMMAND cd ../${OPENCV_BUILD_PATH} && make install -j8
 )
 
+
```
Comment (eggerk) [inline CMakeLists.txt]: rm

File: CMakeLists.txt [hunk 3]
```diff
@@ -82,15 +82,23 @@ ExternalProject_Add(opencv3_src
...
 cs_add_library(${PROJECT_NAME} src/dependency_tracker.cc)
 add_dependencies(${PROJECT_NAME} opencv3_src)
-target_link_libraries(${PROJECT_NAME}
-  ${CATKIN_DEVEL_PREFIX}/lib/libopencv_core${CMAKE_SHARED_LIBRARY_SUFFIX})
+# This can be removed once Ubuntu 14.04 is no longer supported.
+execute_process(COMMAND lsb_release -r OUTPUT_VARIABLE UBUNTU_RELEASE)
```
Comment (eggerk) [inline CMakeLists.txt]: use `-c`? and also add `-s` flag (`-rs` or `-cs`) to get a shorter output

File: CMakeLists.txt [hunk 4]
```diff
@@ -82,15 +82,23 @@ ExternalProject_Add(opencv3_src
...
+if("${UBUNTU_RELEASE}" MATCHES "14.04")
+  target_link_libraries(${PROJECT_NAME}
+    ${CATKIN_DEVEL_PREFIX}/lib/x86_64-linux-gnu/libopencv_core${CMAKE_SHARED_LIBRARY_SUFFIX})
+else()
+  target_link_libraries(${PROJECT_NAME}
```
Comment (eggerk) [inline CMakeLists.txt:98]: what is this library used for?

File: CMakeLists.txt (hunk 1 above)
Comment (fabianbl) [reply CMakeLists.txt]: ok I will add it

File: CMakeLists.txt (hunk 4 above)
Comment (fabianbl) [reply CMakeLists.txt:98]: I think this is just the opencv library (I didn't actually change it, the *.so file is just in a different directory for 14.04)

File: CMakeLists.txt (hunk 1 above)
Comment (fabianbl) [reply CMakeLists.txt:58]: ok will do

**Comment by fabianbl** [comment]
test this please

File: CMakeLists.txt (hunk 4 above)
Comment (eggerk) [reply CMakeLists.txt:98]: I think this library is needed and as a matter of fact, it will be removed per #17

File: CMakeLists.txt [hunk 5]
```diff
@@ -82,15 +82,22 @@ ExternalProject_Add(opencv3_src
...
-target_link_libraries(${PROJECT_NAME}
-  ${CATKIN_DEVEL_PREFIX}/lib/libopencv_core${CMAKE_SHARED_LIBRARY_SUFFIX})
+# This can be removed once Ubuntu 14.04 is no longer supported.
+execute_process(COMMAND lsb_release -rs OUTPUT_VARIABLE UBUNTU_RELEASE)
+if("${UBUNTU_RELEASE}" MATCHES "14.04")
+  target_link_libraries(${PROJECT_NAME}
+    ${CATKIN_DEVEL_PREFIX}/lib/x86_64-linux-gnu/libopencv_core${CMAKE_SHARED_LIBRARY_SUFFIX})
+else()
+  target_link_libraries(${PROJECT_NAME}
+    ${CATKIN_DEVEL_PREFIX}/lib/libopencv_core${CMAKE_SHARED_LIBRARY_SUFFIX})
```
Comment (eggerk) [inline CMakeLists.txt:99]: this seems anyway a bit unsafe and inflexible. Maybe you can link against `opencv_core` instead and not provide the entire library path?

File: CMakeLists.txt [hunk 6]
```diff
@@ -82,15 +82,22 @@ ExternalProject_Add(opencv3_src
...
+  INSTALL_COMMAND cd ../${OPENCV_BUILD_PATH} && make install -j8
 )
 
 cs_add_library(${PROJECT_NAME} src/dependency_tracker.cc)
 add_dependencies(${PROJECT_NAME} opencv3_src)
-target_link_libraries(${PROJECT_NAME}
-  ${CATKIN_DEVEL_PREFIX}/lib/libopencv_core${CMAKE_SHARED_LIBRARY_SUFFIX})
+# This can be removed once Ubuntu 14.04 is no longer supported.
+execute_process(COMMAND lsb_release -rs OUTPUT_VARIABLE UBUNTU_RELEASE)
+if("${UBUNTU_RELEASE}" MATCHES "14.04")
```
Comment (eggerk) [inline CMakeLists.txt:94]: use `STREQUAL`, also you probably need to test if this works on mac

**Comment by fabianbl** [comment]
@eggerk Thanks for the review. I think it to already include your changes, I did it in #19

**Pull request closed without merging**